### PR TITLE
Introduce String.pad_{leading,trailing}/3 and String.trim{,_leading,_trailing}/2

### LINKED ISSUE #4433: Replace String.lstrip/rstrip/ljust/rjust that works on code points on versions that works on graphemes

### PR DESCRIPTION
**Pull request by lexmag**
Closes #4433.

### CODE REVIEW

File: lib/elixir/lib/string.ex [hunk 1]
```diff
@@ -813,6 +813,83 @@ defmodule String do
...
+  end
+
+  def pad_trailing(string, count, [_ | _] = padding)
+      when is_integer(count) and count >= 0 do
+    pad(:trailing, string, count, padding)
+  end
+
+  defp pad(kind, string, count, padding) do
+    string_len = length(string)
+    if string_len >= count do
```
Comment (ericmj) [inline lib/elixir/lib/string.ex]: Move the `if` to a guard?

Comment (ericmj) [inline lib/elixir/lib/string.ex]: nvm, saw that you use the length below. I think the compiler doesn't optimize away the extra `length/1` call?

**Comment by ericmj** [comment]
Looks great. :shipit:

**Comment by whatyouhide** [comment]
Aren't we gonna test `trim_leading` as well?

**Comment by josevalim** [comment]
Looks great.

Just as a reminder, we also need `String.trim/1`, `String.trim_leading/1` and `String.trim_trailing/1` for trailing whitespace. We may want to use more explicit names too. Also, let's not forget to remove the docs for the `*strip` and `*just` functions and leave a `# TODO` so we eventually deprecate them (no warning for now).

**Comment by lexmag** [comment]
Docs, `trim_leading` test and removal of docs for the `*strip` and `*just` are coming. 💛
I think I should have added these as checkboxes initially, when PR was opened, to eliminate the confusion.

**Comment by lexmag** [comment]
This is ready for the final review.

File: lib/elixir/lib/string.ex [hunk 2]
```diff
@@ -750,129 +733,248 @@ defmodule String do
...
 
   @doc ~S"""
-  Returns a new string of length `len` with `subject` right justified and
-  padded with `pad`. If `pad` is not present, it defaults to
-  whitespace. When `len` is less than the length of `subject`, `subject` is
-  returned.
+  Returns a new string padded with a leading filler
+  which is made of elements from the `padding`.
+
+  Passing a string `pagging` is equivalent to passing the graphemes in it.
```
Comment (michalmuskala) [inline lib/elixir/lib/string.ex]: s/pagging/padding/

File: lib/elixir/lib/string.ex [hunk 3]
```diff
@@ -750,129 +733,248 @@ defmodule String do
...
 
   @doc ~S"""
-  Returns a new string of length `len` with `subject` right justified and
-  padded with `pad`. If `pad` is not present, it defaults to
-  whitespace. When `len` is less than the length of `subject`, `subject` is
-  returned.
+  Returns a new string padded with a leading filler
+  which is made of elements from the `padding`.
+
+  Passing a string `padding` is equivalent to passing the graphemes in it.
```
Comment (josevalim) [inline lib/elixir/lib/string.ex]: Let's add a paragraph before this one saying: "Passing a list of strings as `padding` will take one element of the list for every missing entry. If the list is shorter than the number of pads, the padding will start again from the beginning of the list."

**Comment by josevalim** [comment]
:heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:

**Pull request merged**